smolyay/surrogate.py

```diff
@@ -1,546 +1,729 @@
+        the regularization method for determining the coefficients
+    """
+
+    def __init__(self, domain, basis_sets, regularization=None):
+        super().__init__(domain)
+        self._basis_sets = list(basis_sets)
+        self._regularization = None
+        self._index_combinations = None
+        self._coefficients = None
+        self._integration_constant = 0
+        self._needs_terms_constructed = True
+
+        self.regularization = regularization
+
+    @property
+    def regularization(self):
+        """:class:RegularizationMethod: constant of the L2 and L1 term"""
+        return self._regularization
+
+    @regularization.setter
+    def regularization(self, value):
+        if value is not None and not isinstance(value, RegularizationMethod):
+            raise ValueError("Regularization must be a RegularizationMethod")
+        if self.regularization != value:
+            self._regularization = value
+            self._needs_fit = False
+
+    @property
+    def basis_sets(self):
+        """list of BasisFunctionSet: the set of basis functions for the terms."""
+        return self._basis_sets
+
+    def predict(self, X):
         """Evaluate surrogate at a given input.
 
-        Parameter
-        ---------
-        x: list (for dimension > 1), or float (for dimension = 1)
-            Input.
+        Parameters
+        ----------
+        X: array-like with shape (n_samples, num_dimensions) or list of object
+            Points at which the model is evaluated.
 
         Returns
         -------
-        float
+        ndarray of shape (n_samples,) or (n_samples, num_dimensions)
             Surrogate output at x.
 
         Raises
         ------
-        ValueError
+        RuntimeError
             For surrogate to be evaluated, function needs to be trained.
         IndexError
-            Input of the surrogate must be of length of the
-            function's dimensionality.
+            Input must be 2D array with shape (n_samples, num_dimensions).
         ValueError
             Input must lie in domain of surrogate.
+
         """
-        if self._coefficients is None:
-            raise ValueError('Function needs training!')
-        input_shape = numpy.shape(x)
-        x = numpy.array(x, copy=False, ndmin=2)
-        if self.dimension == 1:
-            if len(input_shape) <= 1:
-                # the cast to 2d puts these in wrong order, so transpose
-                x = x.T
-            elif x.shape[-1] > 1:
-                x = x[..., numpy.newaxis]
-        
-        if x.shape[-1] != self.dimension:
-            raise IndexError("Input must match dimension of domain")
-        if self.dimension > 1:
-            oob = any(numpy.any(x[..., i] < self.domain[i][0]) or 
-                    numpy.any(x[..., i] > self.domain[i][1]) 
-                    for i in range(self.dimension))
+        # validate inputs
+        if not self._needs_fit:
```

Comment:
Like here, for example:
```suggestion
        if self._needs_fit:
```